aggregator/follower_flush_mgr.go

```diff
@@ -0,0 +1,350 @@
+		unmarshalErrors:        scope.Counter("unmarshal-errors"),
+		shardNotFound:          scope.Counter("shard-not-found"),
+		resolutionNotFound:     scope.Counter("resolution-not-found"),
+		kvUpdateFlush:          scope.Counter("kv-update-flush"),
+		forcedFlush:            scope.Counter("forced-flush"),
+		flushTimesNotProcessed: scope.Counter("flush-times-not-processed"),
+		flushWindowsNotEnded:   scope.Counter("flush-windows-not-ended"),
+	}
+}
+
+type followerFlushManager struct {
+	sync.RWMutex
+
+	nowFn                 clock.NowFn
+	checkEvery            time.Duration
+	workers               xsync.WorkerPool
+	flushTimesKeyFmt      string
+	flushTimesStore       kv.Store
+	maxNoFlushDuration    time.Duration
+	forcedFlushWindowSize time.Duration
+	logger                xlog.Logger
+	scope                 tally.Scope
+
+	doneCh          <-chan struct{}
+	proto           *schema.ShardSetFlushTimes
+	flushTimesKey   string
+	flushTimesState flushTimesState
+	lastFlushed     time.Time
+	openedAt        time.Time
+	flushTask       *followerFlushTask
+	sleepFn         sleepFn
+	metrics         followerFlushManagerMetrics
+}
+
+func newFollowerFlushManager(
+	doneCh <-chan struct{},
+	opts FlushManagerOptions,
+) roleBasedFlushManager {
+	nowFn := opts.ClockOptions().NowFn()
+	instrumentOpts := opts.InstrumentOptions()
+	scope := instrumentOpts.MetricsScope()
+	mgr := &followerFlushManager{
+		nowFn:                 nowFn,
+		checkEvery:            opts.CheckEvery(),
+		workers:               opts.WorkerPool(),
+		flushTimesKeyFmt:      opts.FlushTimesKeyFmt(),
+		flushTimesStore:       opts.FlushTimesStore(),
+		maxNoFlushDuration:    opts.MaxNoFlushDuration(),
+		forcedFlushWindowSize: opts.ForcedFlushWindowSize(),
+		logger:                instrumentOpts.Logger(),
+		scope:                 scope,
+		doneCh:                doneCh,
+		proto:                 &schema.ShardSetFlushTimes{},
+		flushTimesState:       flushTimesUninitialized,
+		lastFlushed:           nowFn(),
+		sleepFn:               time.Sleep,
+		metrics:               newFollowerFlushManagerMetrics(scope),
+	}
+	mgr.flushTask = &followerFlushTask{mgr: mgr}
+	return mgr
+}
+
+func (mgr *followerFlushManager) Open(shardSetID string) {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	mgr.flushTimesKey = fmt.Sprintf(mgr.flushTimesKeyFmt, shardSetID)
+	mgr.openedAt = mgr.nowFn()
+	go mgr.watchFlushTimes()
+}
+
+// NB(xichen): no actions needed for initializing the follower flush manager.
+func (mgr *followerFlushManager) Init([]*flushBucket) {}
+
+func (mgr *followerFlushManager) Prepare(buckets []*flushBucket) (flushTask, time.Duration) {
+	var (
+		shouldFlush        = false
+		flushersByInterval []flushersGroup
+	)
+	// NB(xichen): a flush is triggered in the following scenarios:
+	// * The flush times persisted in kv have been updated since last flush.
+	// * Sufficient time (a.k.a. maxNoFlushDuration) has elapsed since last flush.
+	now := mgr.nowFn()
+	mgr.Lock()
```

Comment:
Any reason not to `defer mgr.Unlock()` here to avoid a bug being introduced in the future?  Seems we only save on a few extra CPU instructions at the bottom setting values by not using defer here.